[MLIR][AMDGPU] Add OCP FP8 support for new hardware

pcf000 opened the pull request:
Upcoming hardware (gfx12 and some future gfx9) will support the OCP 8-bit float formats for their matrix multiplication intrinsics and conversion operations, retaining existing opcodes and compiler builtins.

This commit adds support for these types to the MLIR wrappers around such operaitons, ensuring that the OCP types aren't used to generate those builtins on hardware that doesn't expect that format and, conversely, to ensure that the pre-OCP formats aren't used on new hardware.

pcf000 committed:
[MLIR][AMDGPU] Add OCP FP8 support to for new hardware

Upcoming hardware (gfx12 and some future gfx9) will support the OCP
8-bit float formats for their matrix multiplication intrinsics and
conversion operations, retaining existing opcodes and compiler builtins.

This commit adds support for these types to the MLIR wrappers around
such operations, ensuring that the OCP types aren't used to generate
those builtins on hardware that doesn't expect that format and,
conversely, to ensure that the pre-OCP formats aren't used on new
hardware.

pcf000 commented:
@krzysz00 @giuseros (I can't add reviewers.)

krzysz00 commented:
@pcf000 There've been downstream changes that fix bugs here, could we add those to this PR

krzysz00 committed:
[MLIR][AMDGPU] After fp8 conversions were lowered to AMDGPU dialect ops,
those operations were not being converted to the LLVM intrinsics they
correspond to because the rewrite patterns were still checking for
gfx940+.

As part of this, factor out tests for type-match isto isNativeFp8()
and isNativeBf8() functions in the AMDGPUToRocdl rewrites.

Also, fix a typo in isGfx940() that caused it to be true for gfx950.

Finally, test all these OCP format conversions by duplicating the
gfx940 tests.

pcf000 committed:
[MLIR][AMDGPU] Clean up and redo after other recent patches here.

kuhar commented on mlir/include/mlir/Dialect/AMDGPU/Utils/Chipset.h:
```diff
@@ -47,6 +47,13 @@ struct Chipset {
   DEFINE_COMP_OPERATOR(>=)
 #undef DEFINE_COMP_OPERATOR
+
+  bool isGfx940() const {
+    return majorVersion == 9 && minorVersion >= 4 && minorVersion < 5;
+  }
+  bool hasOcpFp8() const {
+    return (majorVersion == 9 && minorVersion >= 5) || majorVersion >= 12;
+  }
```
I don't think these belong in the chipset class, they can be defined as free functions where needed.

kuhar commented on mlir/include/mlir/Dialect/AMDGPU/Utils/Chipset.h:
```diff
@@ -47,6 +47,13 @@ struct Chipset {
   DEFINE_COMP_OPERATOR(>=)
 #undef DEFINE_COMP_OPERATOR
+
+  bool isGfx940() const {
+    return majorVersion == 9 && minorVersion >= 4 && minorVersion < 5;
```
The condition is weird, this should be just `minorVersion == 4`. Also, I'd call it something like `isGfx940Series`, since it's not a strict match.

pcf000 replied:
Faulty merge.  It made more sense before steppingVersion, as >=0x40 && < 0x50.

kuhar commented on mlir/lib/Conversion/AMDGPUToROCDL/AMDGPUToROCDL.cpp:
```diff
@@ -454,6 +454,20 @@ static void wmmaPushOutputOperand(ConversionPatternRewriter &rewriter,
 
+/// Return true if `type` is the E5M2 variant of an 8-bit float that is
+/// supported by the `_bf8` instructions on the given `chipset`.
+static bool isNativeBf8(Chipset chipset, Type type) {
+  return (chipset.isGfx940() && type.isFloat8E5M2FNUZ()) ||
+         (chipset.hasOcpFp8() && type.isFloat8E5M2());
+}
+
+/// Return true if `type` is the E4M3FN variant of an 8-bit float that is
+/// supported by the `_fp8` instructions on the given `chipset`.
+static bool isNativeFp8(Chipset chipset, Type type) {
```
nit: `hasNativeF8` to follow the type naming scheme in llvm/mlir?

krzysz00 replied:
... yeah, I can see `hasNaviveF8` or `isNativeF8` if we switched the argument order

pcf000 replied:
This file uses "FP8" extensively, but both as a generic term for Float8 and as the specific for 4.3 (with "BF8" for 5.2).  The two isNative functions are doing the latter.

kuhar commented on mlir/lib/Conversion/ArithToAMDGPU/ArithToAMDGPU.cpp:
```diff
@@ -68,6 +72,15 @@ struct TruncfToFloat16RewritePattern final
 
 } // end namespace
 
+static LogicalResult isSupportedFp8(Type elementType, Chipset chipset) {
```
nit: `supportsF8`?

krzysz00 replied:
I think we've been using Fp8 in this file? But my memory of having written this code ages ago could be hazy

pcf000 replied:
The various ops use "fp8" as do the saturate parameters.  I made this guy isSupportedF8().

kuhar commented on mlir/lib/Conversion/ArithToAMDGPU/ArithToAMDGPU.cpp:
```diff
@@ -68,6 +72,15 @@ struct TruncfToFloat16RewritePattern final
 
+static LogicalResult isSupportedFp8(Type elementType, Chipset chipset) {
+  if (chipset.isGfx940())
+    return success(elementType.isFloat8E5M2FNUZ() ||
+                   elementType.isFloat8E4M3FNUZ());
+  if (chipset.hasOcpFp8())
+    return success(elementType.isFloat8E5M2() || elementType.isFloat8E4M3FN());
```
use `isa<T1, T2>(elementType)`

kuhar commented on mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp:
```diff
@@ -272,14 +272,16 @@ LogicalResult MFMAOp::verify() {
 
   Type sourceBType = getSourceB().getType();
-  if (sourceElem.isFloat8E5M2FNUZ() || sourceElem.isFloat8E4M3FNUZ()) {
+  if (sourceElem.isFloat8E5M2FNUZ() || sourceElem.isFloat8E4M3FNUZ() ||
+      sourceElem.isFloat8E5M2() || sourceElem.isFloat8E4M3FN()) {
```
Also here and in the other conditions below.

It may also make sense to add a helper in some core header that checks if the type is any f8 type.

pcf000 replied:
I invented isFloat(), by analogy with isInteger().  Is that what you had in mind?

pcf000 committed:
[MLIR][AMDGPU] Changes from the review.

kuhar commented on mlir/lib/Conversion/AMDGPUToROCDL/AMDGPUToROCDL.cpp:
```diff
@@ -454,6 +454,20 @@ static void wmmaPushOutputOperand(ConversionPatternRewriter &rewriter,
 
+/// Return true if `type` is the E5M2 variant of an 8-bit float that is
+/// supported by the `_bf8` instructions on the given `chipset`.
+static bool isNativeBf8(Chipset chipset, Type type) {
```
```suggestion
static bool hasNativeBF8(Chipset chipset, Type type) {
```

pcf000 replied:
Since it's a query about the type, not about the chipset, I'm happier with an "is" form.  Since I'm wordy, how about isNativeBf8ForChipset(Type,Chipset)?

kuhar replied:
I don't understand the distinction. The ISA and the supported types are the property of the chipset, no?

pcf000 replied:
Well, yes, but all chipsets have a native BF8 type -- FNUZ for the older ones, OCP for the newer -- so the answer to "has native BF8" is "yes".  Rather, we want to know "is this type the same as the chipset's native BF8 type?"  Also, in a context in which we're doing isa<T1,T2>() and isFloat(), "is" flows better.

kuhar replied:
>  Also, in a context in which we're doing isa<T1,T2>() and isFloat(), "is" flows better.

The issue with chains of `.isType` is that it won't be optimized across multiple types and always performed one by one (with short circuiting). Our casting infra might not do it today, but in order to utilize faster 'group' isa, you have to use it across the codebase in the first place.

kuhar replied:
>  "is this type the same as the chipset's native BF8 type?"
Thanks for the explanation, I see the difference now. Maybe something like this then `isSupportedByNativeBf8(Chipset, Type)`?

If you prefer the current name, could you add a comment that explains what the intention is (similar to how you explained it above)?

krzysz00 replied:
An equivalent way of phrasing it, from where I'm standing, is "is this one of the types that the `_fp8` or `_bf8` instructions on the given `chipset` will be expecting"?

pcf000 replied:
Previous job would have called it "typeIsNativelySupportedBf8ForChipset(Type,Chipset)" -- we liked naming the parameters in the function name -- or "typeIsExpectedBf8ForChipset(Type,Chipset)".  As I wrote those out, I think I like "expected" better than "native" or "supported," but I could be persuaded.

krzysz00 replied:
Hm ... `isImplementedBf8OnChipset()`?

But `Expected` works too

kuhar commented on mlir/include/mlir/IR/Types.h:
```diff
@@ -140,6 +140,9 @@ class Type {
   bool isF64() const;
   bool isF80() const;
   bool isF128() const;
+  /// Return true if this is an float type (with the specified width).
+  bool isFloat() const;
+  bool isFloat(unsigned width) const;
```
nice!

kuhar commented on mlir/lib/Conversion/AMDGPUToROCDL/AMDGPUToROCDL.cpp:
```diff
@@ -454,6 +454,20 @@ static void wmmaPushOutputOperand(ConversionPatternRewriter &rewriter,
+
+/// Return true if `type` is the E4M3FN variant of an 8-bit float that is
+/// supported by the `_fp8` instructions on the given `chipset`.
+static bool isNativeFp8(Chipset chipset, Type type) {
```
```suggestion
static bool hasNativeFp8(Chipset chipset, Type type) {
```

kuhar commented on mlir/lib/Dialect/Tosa/Transforms/TosaValidation.cpp:
```diff
@@ -509,7 +509,9 @@ bool TosaValidation::isValidElementType(Type type) {
     if (profile == TosaProfileEnum::BaseInference)
       return false;
-    return type.isF32() || type.isF16() || type.isBF16();
+    return type.isF32() || type.isF16() || type.isBF16() ||
+           type.isFloat8E4M3FNUZ() || type.isFloat8E5M2FNUZ() ||
+           type.isFloat8E4M3FN() || type.isFloat8E5M2();
```
use isa

pcf000 committed:
[MLIR][AMDGPU] Renaming using suggestions from review.

pcf000 committed:
Use isa<>.

krzysz00 commented:
Pinging this

eric-k256 commented on mlir/lib/Dialect/Tosa/Transforms/TosaValidation.cpp:
```diff
@@ -509,7 +509,8 @@ bool TosaValidation::isValidElementType(Type type) {
   if (isa<FloatType>(type)) {
     if (profile == TosaProfileEnum::BaseInference)
       return false;
-    return type.isF32() || type.isF16() || type.isBF16();
```
The TOSA spec only supports the OCP versions of the FP8 formats, and this pass is verifying adherence to the spec, so we should only add the Float8E4M3FNType and Float8E5M2Type in this location.

eric-k256 commented on mlir/lib/IR/Types.cpp:
```diff
@@ -56,6 +56,15 @@ bool Type::isF64() const { return llvm::isa<Float64Type>(*this); }
 
+bool Type::isFloat() const { return llvm::isa<FloatType>(*this); }
+
+/// Return true if this is an integer type with the specified width.
```
typo integer should be float?

eric-k256 reviewed (commented):
A couple of minor comments on the TOSA part of the change.

mirza-halilcevic commented:
Superseded by #127728, where remaining review comments have been addressed.
